components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts

```diff
@@ -0,0 +1,335 @@
+import { GetCookbooks } from 'app/entities/cookbooks/cookbook.actions';
+import { Cookbook } from 'app/entities/cookbooks/cookbook.model';
+import {
+  allCookbooks,
+  getAllStatus as getAllCookbooksForOrgStatus
+} from 'app/entities/cookbooks/cookbook.selectors';
+import { CreateEnvironment, GetEnvironments } from 'app/entities/environments/environment.action';
+
+const CREATE_TAB_NAME = 'environmentTab';
+
+export class DynamicGrid {
+  id: number;
+  name: string;
+  operator: string;
+  version: string;
+}
+
+@Component({
+  selector: 'app-create-environment-modal',
+  templateUrl: './create-environment-modal.component.html',
+  styleUrls: ['./create-environment-modal.component.scss']
+})
+
+export class CreateEnvironmentModalComponent implements OnInit, OnDestroy {
+
+  @Input() openEvent: EventEmitter<void>;
+  @Input() environmentsList: Environment[] = [];
+  @Input() serverId: string;
+  @Input() orgId: string;
+
+  public visible = false;
+  public creating = false;
+  public conflictError = false;
+  public constraintsTab = false;
+  public detailsTab = true;
+  public isLinear = true;
+  public defaultTab = false;
+  public overrideTab = false;
+  public showdrag = false;
+  public showConstraint = false;
+
+  public detailsFormGroup: FormGroup;
+  public defaultAttrFormGroup: FormGroup;
+  public overrideAttrFormGroup: FormGroup;
+  public constraintKeys: string[] = [];
+  public cookbooks: Cookbook[] = [];
+  public attr_value = '{}';
+  public server: string;
+  public org: string;
+  public per_page = 9;
+  public page = 1;
+  public name_id = '';
+  public jsonString: string;
+  public defaultAttrParseError: boolean;
+  public overrideAttrParseError: boolean;
+  public data: any;
+  public conflictErrorEvent = new EventEmitter<boolean>();
+  public close = new EventEmitter();
+  public constraintArray: Array<DynamicGrid> = [];
+  public items: Environment[] = [];
+  public textareaID: string;
+  public dynamicArray: Array<DynamicGrid> = [];
+  private isDestroyed = new Subject<boolean>();
+
+  constructor(
+    private fb: FormBuilder,
+    private store: Store<NgrxStateAtom>,
+    private telemetryService: TelemetryService
+
+  ) {
+
+    this.detailsFormGroup = this.fb.group({
+      name: ['', [Validators.required, Validators.pattern(Regex.patterns.NON_BLANK)]],
+      description: ['', [Validators.required, Validators.pattern(Regex.patterns.NON_BLANK)]]
+    });
+
+    this.defaultAttrFormGroup = this.fb.group({
+      default: ['{}']
+    });
+
+    this.overrideAttrFormGroup = this.fb.group({
+      override: ['{}', [Validators.required]]
+    });
+
+  }
+
+  ngOnInit(): void {
+    this.openEvent.pipe(takeUntil(this.isDestroyed))
+      .subscribe(() => {
+      this.conflictError = false;
+      this.visible = true;
+      this.items = this.environmentsList;
+      this.showdrag = true;
+      this.showConstraint = true;
+      this.server = this.serverId;
+      this.org = this.orgId;
+    });
+
+    const payload = {
```

Comment:
As mentioned in a previous file, constant objects need a strong type so we know all properties are valid.

Reply:
added type